compaction/compaction.cc

```diff
@@ -1212,18 +1212,57 @@ class cleanup_compaction final : public regular_compaction {
         // optimize set of potentially overlapping ranges by deoverlapping them.
         non_owned_ranges = dht::partition_range::deoverlap(std::move(non_owned_ranges), dht::ring_position_comparator(*_schema));
 
-        // subtract *each* owned range from the partition range of *each* sstable*,
+        // subtract the owned ranges from the partition ranges of all sstables
         // such that we'll be left only with a set of non-owned ranges.
-        for (auto& owned_range : owned_ranges) {
-            dht::partition_range_vector new_non_owned_ranges;
-            for (auto& non_owned_range : non_owned_ranges) {
-                auto ret = non_owned_range.subtract(owned_range, dht::ring_position_comparator(*_schema));
-                new_non_owned_ranges.insert(new_non_owned_ranges.end(), ret.begin(), ret.end());
-                seastar::thread::maybe_yield();
+        auto cmp = dht::ring_position_comparator(*_schema);
+        dht::partition_range_vector new_non_owned_ranges;
+        new_non_owned_ranges.reserve(non_owned_ranges.size() * 2);
+        auto non_owned_range = non_owned_ranges.begin();
+        auto owned_range = owned_ranges.begin();
+        while (non_owned_range != non_owned_ranges.end()) {
+            if (owned_range == owned_ranges.end()) {
+                // We're done with owned_ranges
+                new_non_owned_ranges.emplace_back(std::move(*non_owned_range));
+                ++non_owned_range;
+                continue;
             }
-            non_owned_ranges = std::move(new_non_owned_ranges);
+
+            auto ret = non_owned_range->subtract(*owned_range, cmp);
+            auto size = ret.size();
+            switch (size) {
+            case 0:
+                // current non_owned_range is fully owned, done with it
+                ++non_owned_range;
+                break;
```

Comment:
let's say we have:

non-owned ranges [1, 2], [4, 4]

owned ranges: [2, 2], [4, 5]


on first subtraction, we get range: [1, 1], and advance iterator to non-owned range [4, 4], whereas iterator to owned is still at [2, 2]

on next subtraction, we incorrectly subtract [2, 2] from [4, 4] instead of second owned range which is [4, 5].

in case 0, shouldn't we make that iterator to owned range is advanced until non-owned is before or contained in the current owned range?

the direction of implementation looks good to me (other than the possible problem above), but it was hard for me to grasp all details at this time of the night :-) I'll probably forget about it tomorrow. I think this is a good candidate for unit test, we could add a public static function in cleanup_compaction:

dht::partition_range_vector
    get_ranges_for_invalidation(partition_range_vector owned_ranges, partition_range_vector non_owned_ranges)

then use it in the unit test and cleanup itself.

another thing that came to my mind:
we can probably do some renaming to make the procedure easier to follow; I had a problem with the names being too similar. 
non_owned_ranges -> candidate_ranges
new_non_owned_ranges -> disowned_ranges or ranges_for_invalidation

Reply:
> let's say we have:
> 
> non-owned ranges [1, 2], [4, 4]
> 
> owned ranges: [2, 2], [4, 5]
> 
> on first subtraction, we get range: [1, 1], and advance iterator to non-owned range [4, 4], whereas iterator to owned is still at [2, 2]
> 
> on next subtraction, we incorrectly subtract [2, 2] from [4, 4] instead of second owned range which is [4, 5].

We shouldn't.  On next iteration, [2, 2] and [4, 4] do not intersect, hence [4, 4].subtract([2, 2]) = [4, 4],
so we fall into the size==1 case below, and since cmp(2, 4) < 0 we will set `*non_owned_range = std::move(ret[0]);` that will keep it the same, and increment `++owned_range`.

> 
> in case 0, shouldn't we make that iterator to owned range is advanced until non-owned is before or contained in the current owned range?
> 
> the direction of implementation looks good to me (other than the possible problem above), but it was hard for me to grasp all details at this time of the night :-) I'll probably forget about it tomorrow. I think this is a good candidate for unit test, we could add a public static function in cleanup_compaction:
> 
> dht::partition_range_vector get_ranges_for_invalidation(partition_range_vector owned_ranges, partition_range_vector non_owned_ranges)
> 
> then use it in the unit test and cleanup itself.

We can add a unit test that go over the results and make sure they are contained in full in the original non_owned_ranges and they don't intersect with owned_ranges.

> 
> another thing that came to my mind: we can probably do some renaming to make the procedure easier to follow; I had a problem with the names being too similar. non_owned_ranges -> candidate_ranges new_non_owned_ranges -> disowned_ranges or ranges_for_invalidation

no problem.

Reply:
thanks for the explanation.